os/pm/pm_ioctl.c

```diff
@@ -0,0 +1,143 @@
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <assert.h>
+#include <tinyara/pm/pm.h>
+#include <tinyara/clock.h>
+#include <tinyara/irq.h>
+
+#include "pm.h"
+
+#ifdef CONFIG_PM
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: pm_ioctl
+ *
+ * Description:
+ *   This function can be called from applications running on TizenRT to
+ *   request/enable/disable power state transitions. This function conveys
+ *   the request from the application to the PM module, which thereafter
+ *   checks the feasibility of the requests. There is no returned value.
+ *
+ *   The power management state is not automatically changed, however.
+ *   up_idlepm() must be called in order to make the state change.
+ *
+ *   These two steps are separated because the application is not supposed
+ *   to invoke power state changes directly, it can only request/suggest.
+ *
+ * Input Parameters:
+ *   domain - the PM domain to check
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void pm_ioctl(enum pm_ioctl_cmds cmd)
+{
+	switch(cmd) {
+		case PM_IOC_NOCMD:
+			dbg("Nothing to do\n");
+			break;
+		case PM_IOC_STAY:
+			//Get the current domain
+			//perform pm_stay to block the domain in its current PM state
+			//pm_stay(<DOMAIN>, pm_querystate(<DOMAIN>));
+			break;
+		case PM_IOC_RELAX:
+			//Get the current domain
+			//perform pm_relax to unblock the domain's state transitions
+			//pm_relax(<DOMAIN>, pm_querystate(<DOMAIN>));
+			break;
+		/* Note that in the above 2 IOCTL calls, the application does not provide the state to stay/relax.
+		 * The application cannot anyways know the pm_domain of the current thread.
+		 * The thread will be extracted during the processing of the IOCTL call itself.
+		 * The state to stay in is the current state itself, and the state to be relaxed from will also
+		 * be the current state itself. This is based on the assumption that if an application is requesting
+		 * a stay, then the current state is capable enough for the application to do its work, hence we will
+		 * stay in the current request only. If a stay was invoked in the current PM state, then it should be
+		 * relaxed from the same state too.
+		 * Open to discussion and consideration
+		 */
+		case PM_IOC_SLEEP:
+			//Get the current domain
+			/* Perform any checks neccessary for doing the next step. As per current code in idle.c, some
+			 * preliminary steps are performed in STANDBY state, and then state changes to SLEEP.
+			 */
+			/* Change the recommended state for the current domain to PM_SLEEP
```

Comment:
Other than changing the recommended state directly, is there any other operation need to be done to handle such change? Because there might be possibility that the pdom->recommended is being changed at the same timing, I'm thinking that we might have to add some lock access for such variables to prevent from racing condition?
As this pm_ioctl.c will be handled independently from the IDLE thread.

Reply:
Multiple apps may request for sleep at the same time.

But if we are considering only 1 core right now, only 1 application can be requesting sleep at a time. Later, if the app is switched, then other app can request sleep. But by this time, the first app's request has been processed as I mentioned in Line 135

In a dual core scenario, two apps may request PM_IOC_SLEEP at the same time. In this case we can have a worse condition where we cannot determine whose SLEEP request gets granted first.

We can clarify such critical points as we go. Please point out if you think I have missed anything.

Reply:
I think this discussion thread has similar meaning to the below one, if we are doing state transition in a non-natural way, we can modify the logic for up_idlepm(), to prevent conflict from 2 points:
1) pdom->recommended being modified by pm_ioctl()
2) pdom->recommended being modified by pm_checkstate()
Please correct if my understanding is incorrect

Reply:
Yes, this needs to be taken care of. Also any other such critical synchronisation points will also need to be identified and taken care of.
We may see many cases when the application directly tries to invoke sleep. 
PM_IOC_STAY and RELAX should be simpler.